[FIX] call pthread_setname_np() from the thread to be renamed

RomainNaour opened the pull request:
pthread_setname_np() must be used from the thread to be renamed
otherwise the new name will not be accepted when the openPOWERLINK
application is executed from a restricted user (i.e not root).

This is explained by the pthread_setname_np() implementation [1].

If we try to rename the thread calling pthread_setname_np(), then
"prctl (PR_SET_NAME, name)" will be used and the new name will be accepted.

But if we try to rename another thread, the PROCFS will be used
(/proc/self/task/<pid>/comm) and for some reason "comm" can't be written by
the user.

Make sure to rename a thread by calling pthread_setname_np() from the thread
to be renamed.

[1] https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/pthread_setname.c;h=6c5e1d6e485eaffed227156b6ccd75e54cc34a03;hb=fdfc9260b61d3d72541f181$

RomainNaour commented:
Hi,
This is probably a right access issue to the procfs on my system since we drop most of capabilities when the application start.
It's not clear to me why the write access to /proc/self/task/<pid>/comm fails, it seems legit to rename the thread after phtread_create() (see the exemple in man pthread_setname_np).
Thought ?

banickn commented:
Hi Romain,

I can't really reproduce the issue with your pull request changes because we always run the application as sudo due to several parts that require it to do so.

As you wrote, it probably makes sense to check if the process is accessible without root.
Also, the renaming of the threads is mostly a convenience feature for easier debugging. Maybe you don't actually need that at all in your use case?

RomainNaour commented:
Hi Norman,

The real-time/openpowerlink application I use, is not executed by root user for "security" reasons.
Actually, the only required capabilities are cap_sys_nice and cap_ipc_lock.
Maybe it would be great to extend the demos applications to take into account this use case.

Threads renaming is quite useful when the application is monitored with htop for example.
But ok it's not something very important.

I'm able to reproduce this issue on my desktop with a minimal piece of code and after some investigation, it seems to be a kernel bug [1]. I'll try to backport this patch in my kernel and do some test.
I'm pretty sure it will fix my issue, so I'm closing the PR right now ;-)

Best regards,
Romain

[1] http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=1b3044e39a89cb1d4d5313da477e8dfea2b5232d

RomainNaour commented:
Hello,

I confirm that's was an kernel issue fixed by [1].

Best regards,
Romain

[1] http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=1b3044e39a89cb1d4d5313da477e8dfea2b5232d

seissw commented:
Hi Romain,

Does that mean that your changes should be applied and should work (when the kernel bug is fixed)?
Or are the changes not required at all and the PR is closed?

Best regards,
Wolfgang

RomainNaour commented:
Hi Wolfgang,

My changes works with and without the kernel patch but since it's a kernel bug I believe that's the kernel that must be patched, that's why I closed the PR.
This change shouldn't be applied in the openPOWERLINK stack.

I reverted it in my stack and added the upstream patch (from 4.7) to my kernel.

But if users care about threads naming, we should recommend to backport the patch in their kernel.
Maybe the kernel patch will be backported to the latest stable kernel... It's not the case for the moment.

Best regards,
Romain
